clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -5984,10 +5984,17 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
                     options::OPT_fno_gnu_inline_asm, true))
     CmdArgs.push_back("-fno-gnu-inline-asm");
 
+  bool EnableSYCLEarlyOptimizations =
+      Args.hasFlag(options::OPT_fsycl_early_optimizations,
+                   options::OPT_fno_sycl_early_optimizations,
+                   Triple.getSubArch() != llvm::Triple::SPIRSubArch_fpga);
+
   // Enable vectorization per default according to the optimization level
   // selected. For optimization levels that want vectorization we use the alias
   // option to simplify the hasFlag logic.
   bool EnableVec = shouldEnableVectorizerAtOLevel(Args, false);
+  if (UseSYCLTriple && EnableSYCLEarlyOptimizations)
+    EnableVec = false; // But disable vectorization for SYCL device code
```

Comment:
Am I right to conclude that vectorization still won't be enabled `if (UseSYCLTriple && !EnableSYCLEarlyOptimizations)`?

Reply:
vectorization will be disabled by default in SYCL mode if optimizations are enabled. 
It can be enabled with `-fvectorize` option.

Reply:
Sorry for obscuring my own question. I meant, what is supposed to happen if we're in SYCL mode **without** early optimizations enabled? Should this code still disable vectorization by default?

Reply:
What "vectorization" do you mean? Back-end compiler should be able to vectorize if HW natively supports vector operations.

Reply:
I think the puzzle has come together for me now, thanks!